src/operations/find.ts

```diff
@@ -89,169 +92,242 @@ export class FindOperation extends CommandOperation<FindOptions, Document> {
-      if (projection && !Buffer.isBuffer(projection) && Array.isArray(projection)) {
-        projection = projection.length
-          ? projection.reduce((result, field) => {
-              result[field] = 1;
-              return result;
-            }, {})
-          : { _id: 1 };
+    if (serverWireVersion < 4) {
+      if (this.readConcern && this.readConcern.level !== 'local') {
+        callback(
+          new MongoError(
+            `server find command does not support a readConcern level of ${this.readConcern.level}`
+          )
+        );
+
+        return;
       }
 
-      findCommand.fields = projection;
+      const findCommand = makeLegacyFindCommand(this.ns, this.filter, options);
+      if (isSharded(server) && this.readPreference) {
+        findCommand.$readPreference = this.readPreference.toJSON();
+      }
+
+      server.query(
+        this.ns,
+        findCommand,
+        {
+          ...this.options,
+          ...this.bsonOptions,
+          documentsReturnedIn: 'firstBatch',
+          readPreference: this.readPreference
+        },
+        callback
+      );
+
+      return;
     }
 
-    if (options.hint) {
-      findCommand.hint = normalizeHintField(options.hint);
+    let findCommand = makeFindCommand(this.ns, this.filter, options);
+    if (this.explain) {
+      findCommand = decorateWithExplain(findCommand, this.explain);
     }
 
-    if (typeof options.skip === 'number') {
-      findCommand.skip = options.skip;
+    server.command(
+      this.ns.toString(),
+      findCommand,
+      {
+        fullResult: !!this.fullResponse,
+        ...this.options,
+        ...this.bsonOptions,
+        documentsReturnedIn: 'firstBatch'
+      },
+      callback
+    );
+  }
+}
+
+function makeFindCommand(ns: MongoDBNamespace, filter: Document, options: FindOptions): Document {
+  const findCommand: Document = {
+    find: ns.collection,
+    filter
+  };
+
+  if (options.sort) {
+    findCommand.sort = formatSort(options.sort);
+  }
+
+  if (options.projection) {
+    let projection = options.projection;
+    if (projection && !Buffer.isBuffer(projection) && Array.isArray(projection)) {
+      projection = projection.length
+        ? projection.reduce((result, field) => {
+            result[field] = 1;
+            return result;
+          }, {})
+        : { _id: 1 };
     }
 
-    if (typeof options.limit === 'number') {
-      if (options.limit < 0 && maxWireVersion(server) >= 4) {
-        findCommand.limit = Math.abs(options.limit);
-        findCommand.singleBatch = true;
-      } else {
-        findCommand.limit = options.limit;
-      }
+    findCommand.projection = projection;
+  }
+
+  if (options.hint) {
+    findCommand.hint = normalizeHintField(options.hint);
+  }
+
+  if (typeof options.skip === 'number') {
+    findCommand.skip = options.skip;
+  }
+
+  if (typeof options.limit === 'number') {
+    if (options.limit < 0) {
+      findCommand.limit = Math.abs(options.limit);
```

Comment:
```suggestion
      findCommand.limit = Math.abs(options.limit);
```

Nit: using the `-` operator is probably faster than Math.abs when we know something is negative, but also probably a fairly negligible improvement. Also applies below in the `options.batchSize < 0` condition.

Reply:
fixed. I left two `Math.abs` in there where I felt it improved readability (`Math.abs(batchSize) < Math.abs(limit)` looks better than `-batchSize < -limit` IMO)